add TopicGroup component that makes possible to group simple topics into subtopics

AndreySV opened the pull request:
additionaly this commit makes that topics are automatically unsubscribed/subscribed
after rename. Before it was done only after connect and reconnect was
required to subscribed renamed topic. Now all is done automatically.

linsong commented on TopicGroup.sedona:
```diff
@@ -0,0 +1,147 @@
+//
+// Copyright 2018 Andrey Skvortsov <andrej.skvortzov@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
in this repository, I choose to use MIT license since it seems more open. To make it consistent, all code files, except 3rd party codes,  should follow the same license. do you mind to remove this license comments here ?

AndreySV replied:
BTW Apache 2 license is as open as MIT, but has patents protection for original authors. It shouldn't be a problem here, but who knows... 

IMHO it's important to have license statement with copyright holders in source file, so it's kept in source file even for permisive license like MIT in case the code is merged into other project. For consistency with all existed source files in the repo I can remove license statement from the header.

linsong commented on MessageCommon.sedona:
```diff
@@ -0,0 +1,144 @@
+  // Fields
+  ////////////////////////////////////////////////////////////////
+  define int fullTopicNameLen = 400
+  protected inline Str(fullTopicNameLen) prefixName = ""
```
for every MessageCommon instance, there will be 3 Buf variables that 400(fullTopicNameLen) bytes, that will be ~1.2KB. maybe we can decrease the memory usage somehow, for example, calculate them on the fly if possible.

AndreySV replied:
I don't think it's a big problem honestly. Especially because this kit is targeting Linux-based systems, where memory requirements are already pretty high in compare to more embedded stuff like cortex-m. I increased default 64 length for topic name, because if user wants subscribe single topic. It easier to just use single MessageInt component with a full topic path instead of build up whole hierarchy of TopicGroup component. Probably for most cases 64 characters will be enough.

I'm not sure that it's possible to decrease memory usage on the fly in sedona. 
And thanks for fixing my implementation. It's nice to get rid of lastSubscribedTopic. That brings a lot of memory saving.

linsong commented on MessageCommon.sedona:
```diff
@@ -0,0 +1,144 @@
+  **
+  ** reset current message's local state
+  **
+  internal virtual void reset() {
```
from reset to unsubscribe methods are pure virtual interface methods here, recommend to use 'abstract' instead of 'virtual' to  make this more obvious.

AndreySV replied:
Good point.
BTW I see you've added internal attributes to a lot of methods (I've followed your convention in my change). But this brings a lot of warnings about unused methods.

This is log from AndreySV-master build:
```[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/Message.sedona:49:3: internal method 'communityMQTT::Message.publish' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/Message.sedona:97:3: internal method 'communityMQTT::Message.subscribe' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/Message.sedona:158:3: internal method 'communityMQTT::Message.onTopicChanged' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/Message.sedona:174:3: internal method 'communityMQTT::Message.reset' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/TopicGroup.sedona:27:3: internal method 'communityMQTT::TopicGroup.reset' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/TopicGroup.sedona:43:3: internal method 'communityMQTT::TopicGroup.publish' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/TopicGroup.sedona:59:3: internal method 'communityMQTT::TopicGroup.subscribe' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/TopicGroup.sedona:75:3: internal method 'communityMQTT::TopicGroup.unsubscribe' is not used
[WARNING] /home/travis/build/linsong/sedona-mqtt/sedona/src/kits/communityMQTT/test/MQTTServiceTest.sedona:1:1: internal class 'communityMQTT::MQTTServiceTest' is not used
  WriteKit [/home/travis/build/linsong/sedona-mqtt/sedona/build/kits/communityMQTT/communityMQTT-92d307a6-1.2.29.kit]
  WriteManifest [/home/travis/build/linsong/sedona-mqtt/sedona/build/manifests/communityMQTT/communityMQTT-92d307a6.xml]```

Wouldn't we better get rid of them?

linsong commented:
Hi Andrey, 

  Sorry, I just submit by mistake. The commit seems good, but I have several comments: 

    * the memory usage per Message is a bit big, I think we can make it smaller
    * to make things more flexible, we can give user the power to use TaskGroup in nested way
    
  I have made some changes locally about above 2 comments, and it works. Before I can push to your fork branch, you need to [enable the permission](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ )
  
  If you agree, I would like to invite you to be a contributor of this repository, then I create my changes as a new pull request, we can collaborate easier.

  Thanks for your great work!

Cheers,
Vincent

AndreySV commented:
Thanks for reducing memory consumption. TopicGroups could be used in a nested way in my implementation already. This was handled by 'updatePrefix' and 'propagatePrefixToChildren' methods.

"Allow edits from maintainers." option is already on for the pull request. So you should be able to push your changes.

I'd be glad to be a contributor of the repo. Let's do that.

I've done some local testing of AndreySV-master branch.  
If Publisher/Subscriber is removed, then 'UNSUBSCRIBE: handle is null' messages appear in the log. 
Because worker is stopped before unsubscribe methods are called. It's ok, but I'd avoid these messages.

```
diff --git a/Subscriber.sedona b/Subscriber.sedona
index d738f37..37be3c8 100644
--- a/Subscriber.sedona
+++ b/Subscriber.sedona
@@ -11,7 +11,8 @@ class Subscriber extends Worker
   override int childEvent(int eType, Component child)
   {
     if (eType == Component.REMOVED && child.type.is(MessageCommon.type)) {
-      ((MessageCommon)child).unsubscribe(handle);
+      if (handle != null)
+        ((MessageCommon)child).unsubscribe(handle);
     }
     return 0
   }
diff --git a/TopicGroup.sedona b/TopicGroup.sedona
index 76ad9ff..5a84bfc 100644
--- a/TopicGroup.sedona
+++ b/TopicGroup.sedona
@@ -111,7 +111,9 @@ class TopicGroup extends MessageCommon
       return 0
 
     if (eType == Component.REMOVED && child.type.is(MessageCommon.type)) {
-      ((MessageCommon) child).unsubscribe(getHandle());
+      Obj handle = getHandle()
+      if (handle != null)
+        ((MessageCommon) child).unsubscribe(handle);
     }
     return 0
   }
```

linsong commented:
I created a [new pull request](https://github.com/linsong/sedona-mqtt/pull/4) based on this one, but with several improvements. So I just close this one and let's add comment in the new pull request, thanks.
